src/mbgl/gl/drawable.hpp

```diff
@@ -0,0 +1,124 @@
+    static constexpr gl::BufferID indexBuffer = 0;
+    static constexpr std::size_t primitiveSize = 0;
+    std::vector<Segment> segments;
+};
+
+template <class Mode, class Vertex, class...Args>
+auto Unindexed(const VertexBuffer<Vertex>& vertexBuffer,
+               Args&&... modeArguments) {
+    return UnindexedVertices<Vertex>(
+        Mode { std::forward<Args>(modeArguments)... },
+        vertexBuffer);
+}
+
+template <class Vertex, class Primitive>
+class SegmentedVertices {
+public:
+    using VertexType = Vertex;
+
+    SegmentedVertices(Mode mode_,
+                      const VertexBuffer<Vertex>& vertexBuffer_,
+                      const IndexBuffer<Primitive>& indexBuffer_,
+                      const std::vector<Segment>& segments_)
+        : mode(std::move(mode_)),
+          vertexBuffer(vertexBuffer_.buffer),
+          indexBuffer(indexBuffer_.buffer),
+          segments(segments_) {}
+
+    Mode mode;
+    gl::BufferID vertexBuffer;
+    static constexpr std::size_t vertexSize = sizeof(Vertex);
+    gl::BufferID indexBuffer;
+    static constexpr std::size_t primitiveSize = sizeof(Primitive);
+    const std::vector<Segment>& segments;
+};
+
+template <class Mode, class Vertex, class Primitive, class...Args>
+auto Segmented(const VertexBuffer<Vertex>& vertexBuffer,
+               const IndexBuffer<Primitive>& indexBuffer,
+               const std::vector<Segment>& segments,
+               Args&&... modeArguments) {
+    static_assert(std::is_same<typename Primitive::Mode, Mode>::value, "primitive mode mismatch");
+    return SegmentedVertices<Vertex, Primitive>(
+        Mode { std::forward<Args>(modeArguments)... },
+        vertexBuffer,
+        indexBuffer,
+        segments);
+}
+
+class Drawable {
+public:
+    template <class Shader, class Subject>
+    Drawable(Depth depth_,
+             Stencil stencil_,
+             Color color_,
+             Shader& shader,
+             typename Shader::UniformsType::Values&& uniformValues,
+             const Subject& subject)
+        : mode(subject.mode),
+          depth(std::move(depth_)),
+          stencil(std::move(stencil_)),
+          color(std::move(color_)),
+          program(shader.getID()),
+          vertexBuffer(subject.vertexBuffer),
+          vertexSize(subject.vertexSize),
+          indexBuffer(subject.indexBuffer),
+          primitiveSize(subject.primitiveSize),
+          segments(subject.segments),
+          bindUniforms(Shader::UniformsType::binder(shader.uniformsState, std::move(uniformValues))),
+          attributeBindings(AttributeBindings<Shader, typename Subject::VertexType>()(shader))
+    {
+        static_assert(std::is_standard_layout<typename Subject::VertexType>::value, "vertex type must use standard layout");
+        static_assert(std::is_same<typename Shader::VertexType, typename Subject::VertexType>::value, "vertex type mismatch");
+    }
+
+    Mode mode;
+    Depth depth;
+    Stencil stencil;
+    Color color;
```

Comment:
I chose `Color` for parallelism with `Depth` and `Stencil` -- each holds the configuration that affects the respective GL buffer attachment of that type. For `Color`, that's both blend settings and the color mask.